doc/contributing/coding-conventions.chapter.md

```diff
@@ -181,6 +181,21 @@
   rev = "${version}";
   ```
 
+- Filling lists condionally _should_ be done with `lib.optional(s)` instead of using `if cond then [ ... ] else null`.
+
+  ```nix
+  buildInputs = lib.optional stdenv.isDarwin iconv;
+  ```
+
+  instead of
+
+  ```nix
+  buildInputs = if stdenv.isDarw then [ iconv ] else null;
+  ```
+
+  An exception can be made when fixing a important bug without triggering a mass rebuild.
+  If this is done a follow up pull request _should_ be created to change the code to `lib.optional(s)`.
```

Comment:
Why would that trigger a rebuild? What do you have in mind?
Furthermore, I guess this is the case for all coding conventions, so I'm not sure that this paragraph is relevant.

Reply:
Re: why — having an `mkDerivation` parameter present but equal to empty list is not the same as not having it at all, and theoretically the builder can observe it (empty vs. unset environment variable).

Reply:
Indeed, its true when defaulting to `null`, not `[]`.

Reply:
> Why would that trigger a rebuild? What do you have in mind?

lib.optional returns [] instead of null which triggers a rebuild.

Reply:
Of course it evaluates to [].

Reply:
Maybe «An exception can be made» should be «The latter can be used»?

I often forget how `null` is handled (and I think it might have changed), so a wording reminding which way actually avoids rebuilds would make it clear why `null` is even mentioned.

Reply:
Which later? There is no previous sentence which can be referenced, just the example.

Reply:
Well, the example puts two styles into opposition, but you are right that more clarity is better. «As an exception, an explicit conditional expression with `null` can be used when …»?